becquerel/core/autocal.py

```diff
@@ -0,0 +1,267 @@
+"""Automatic calibration by associating peaks with energies."""
+
+from __future__ import print_function
+from itertools import combinations
+import matplotlib.pyplot as plt
+import numpy as np
+from .peakfinder import PeakFinder
+from .energycal import LinearEnergyCal
+
+
+MAJOR_BACKGROUND_LINES = [
+    1460.82,
+    2614.51,
+    1620.50,
+    911.20,
+    338.32,
+    238.63,
+    2204.06,
+    1764.49,
+    1377.67,
+    1238.12,
+    1120.29,
+    768.36,
+    609.32,
+    351.93,
+    295.22,
+    242.00,
+    186.21,
+]
+
+
+def fit_gain(channels, snrs, energies):
+    """Calculate the mean gain to relate channels to energies.
+
+    Assume that the peak channel error scales like fwhm/snr,
+    which is proportional to sqrt(x)/snr.
+
+    Minimize:
+        F = sum(snr^2 * (x * g - y)^2 / x)
+        dF/dg = sum(2 * snr^2 * x * (x * g - y) / x)
+              = 2 * sum(snr^2 * (x * g - y))
+              = 2 * (g * sum(snr^2 * x) - sum(snr^2 * y))
+              = 0
+        g = sum(snr^2 * y) / sum(snr^2 * x)
+
+    """
+    assert len(channels) == len(energies)
+    assert len(channels) == len(snrs)
+    x = np.asarray(channels)
+    s = np.asarray(snrs)
+    y = np.asarray(energies)
+    S2X = (s**2 * x).sum()
+    S2Y = (s**2 * y).sum()
+    gain = S2Y / S2X
+    return gain
+
+
+def fom_gain(channels, snrs, energies):
+    """Calculate a leave-one-out cross-validation figure of merit.
+
+    Assume that the peak channel error scales like fwhm/snr,
+    which is proportional to sqrt(x)/snr. Then the figure of merit should be:
+
+    F = sum(snr^2 * (x * g - y)^2 / x)
+
+    s2x = sum(snr^2 x)
+    s2y = sum(snr^2 y)
+
+    Least-square gain when excluding channel j:
+    g_j = sum_(i!=j)(snr_i^2 y_i) / sum_(i!=j)(snr_i^2 x_i)
+        = (s2y - snr_j^2 y_j)/(s2x - snr_j^2 x_j)
+
+    Figure of merit when excluding channel j:
+    F_j = snr_j^2 (g_j x_j - y_j)^2 / x_j
+
+    Leave-one-out cross validation FOM:
+    F_LOO = sum_j(F_j) / N
+    = sum_j(snr_j^2 (g_j x_j - y_j)^2 / x_j) / N
+    = sum_j(snr_j^2 (s2y x_j - s2x y_j)^2 / (s2x - snr_j^2 x_j)^2 / x_j) / N
+
+    """
+    assert len(channels) == len(energies)
+    assert len(channels) == len(snrs)
+    x = np.asarray(channels)
+    s = np.asarray(snrs)
+    y = np.asarray(energies)
+    N = len(channels)
+    S2X = (s**2 * x).sum()
+    S2Y = (s**2 * y).sum()
+    squared_errs = s**2 * (S2Y * x - S2X * y)**2 / (S2X - s**2 * x)**2 / x
+    return squared_errs.sum() / N / (N - 1)
+
+
+def find_best_gain(
+        channels, snrs, required_energies, optional=(),
+        gain_range=(1e-3, 1e3), de_max=10., verbose=False):
+    """Find the gain that gives the best match of peaks to energies."""
+    assert len(channels) == len(snrs)
+    assert len(channels) >= 2
+    assert len(required_energies) >= 2
+    assert len(channels) >= len(required_energies)
```

Comment:
@markbandstra can you change these from `assert` to raise an `AutoCalibrationError`?